app/test/features/chat_ng/widgets/chat_messages_list_test.dart

```diff
@@ -0,0 +1,406 @@
+        messageList: List.generate(20, (index) => 'msg$index'),
+        loading: const ChatRoomLoadingState.loaded(),
+      );
+
+      final mockTimelineStream = MockTimelineStream();
+      when(
+        () => mockTimelineStream.markAsRead(any()),
+      ).thenAnswer((_) async => false);
+
+      await tester.pumpProviderWidget(
+        overrides: [
+          chatMessagesStateProvider(
+            testRoomId,
+          ).overrideWith((ref) => MockChatRoomMessagesNotifier(messagesState)),
+          chat
+              .timelineStreamProvider(testRoomId)
+              .overrideWith((ref) => Future.value(mockTimelineStream)),
+          animatedListChatMessagesProvider(
+            testRoomId,
+          ).overrideWith((ref) => GlobalKey<AnimatedListState>()),
+          isActiveProvider(LabsFeature.chatUnread).overrideWith((ref) => false),
+          chat.hasUnreadMessages(testRoomId).overrideWith((ref) => true),
+        ],
+        child: ChatMessages(roomId: testRoomId),
+      );
+
+      final chatMessagesWidget = find.byType(ChatMessages);
+      final state = tester.state(chatMessagesWidget) as dynamic;
+
+      await state.onScroll();
+
+      await tester.pump(const Duration(milliseconds: 300));
+
+      // verifymarkAsRead was called
+      verify(() => mockTimelineStream.markAsRead(true)).called(1);
+    });
+
+    testWidgets('handles message insertion animation correctly', (
+      tester,
+    ) async {
+      final initialState = ChatRoomState(
+        messageList: List.generate(3, (index) => 'msg$index'),
+        loading: const ChatRoomLoadingState.loaded(),
+      );
+
+      // Create a real notifier that we can update
+      final notifier = ChatRoomMessagesNotifier(
+        roomId: testRoomId,
+        ref: MockRef(),
+      );
+      notifier.state = initialState;
+
+      final animatedListKey = GlobalKey<AnimatedListState>();
+
+      await tester.pumpProviderWidget(
+        overrides: [
+          chatMessagesStateProvider(testRoomId).overrideWith((ref) => notifier),
+          chat
+              .timelineStreamProvider(testRoomId)
+              .overrideWith((ref) => Future.value(MockTimelineStream())),
+          animatedListChatMessagesProvider(
+            testRoomId,
+          ).overrideWith((ref) => animatedListKey),
+        ],
+        child: ChatMessages(roomId: testRoomId),
+      );
+
+      expect(animatedListKey.currentState?.widget.initialItemCount, equals(3));
+
+      // Verify initial message order
+      expect(notifier.state.messageList, equals(['msg0', 'msg1', 'msg2']));
+
+      // add new message
+      final updatedState = initialState.copyWith(
+        messageList: [...initialState.messageList, 'newMessage'],
+      );
+      notifier.state = updatedState;
+
+      await tester.pump();
+      await tester.pump(const Duration(milliseconds: 300));
+
+      // verify the list was updated
+      expect(animatedListKey.currentState?.widget.initialItemCount, equals(4));
```

Comment:
this doesn't actually check for what the comment says. it is checking only whether the incoming parameter was updated, not whether the actual list items updated. I am not sure we can actually check for the items of the state but this tests creates a false sense of security that will make debugging harder as it would show a false negative about the actual state of the system.

In such a case I'd rather have the diffing test extended with a mocked animatedList that is checking for the right methods to be called. In tandem they can be fairly accurate then.